rfcs/0010-settings-vs-components.md

```diff
@@ -0,0 +1,66 @@
+# Settings design
+
+Status: Draft
+
+Authors:
+
+* [David Calavera](https://github.com/calavera)
+
+## Motivation
+
+SlateDB is an embedded storage engine built on top of object storage. Since it's designed to be used in many different environments, SlateDB exposes a number of configuration options to use fine-tune the engine for the environment it's running in. It also exposes several components that users can swap with their own implementations to customize the engine, for example the `object_store` implementation, or the `clock` implementation.
+
+These configuration options and components are somehow intertwined. Some components are included in the `DbOptions` struct and its related structs. Others are passed as parameters to the `Db::open` method.
+
+This coupling makes it hard to understand where all the components are, what they do, and how they fit together with the configuration options. As an example of this coupling, you can see that the `DbOptions` struct has some fields marked with the macro `#[serde(skip)]` to avoid serializing them. This is because these fields are not part of the settings that a user can change, they are components that are internally used by the engine. One could make the argument that all components should be part of the `DbOptions` struct, but this would make the API more complex and harder to understand.
+
+## Goals
+
+This RFC aims to:
+
+* Separate the concerns of configuration options and components.
+* Make it easier to introduce new components without breaking existing public APIs.
+
+## Proposal
+
+### Separate the concerns of configuration options and components
+
+To make it easier to understand how SlateDB is configured, I propose to separate the concerns of configuration options, or settings, and components:
+
+- Settings are options that a user can tweak to customize the engine for their use case
+- Components are the parts of the database that are responsible for performing the work.
+
+With that in mind, I propose to create a type alias for settings called `Settings` that will be used instead of the `DbOptions` struct. This struct will be cleaned from the components that are not settings. Currently, there are 3 components that are not settings:
+
+- `block_cache`: The block cache to use.
+- `clock`: The clock to use.
+- `gc_runtime`: The runtime to use for garbage collection.
```

Comment:
I think `compaction_scheduler` and `compaction_runtime` might also be components. They're nested in the `compactor_options` field of `DbOptions`.